### PR TITLE
[VQueues] Concrete implementation of invoker throttling

### LINKED ISSUE #4651: Incoming breakage in a future Rust version
I have discovered your repo via a github search. Some of your code currently compiles only due to a bug in the Rust standard library. Once the bug is fixed in a future version of Rust, your code will stop compiling.

The issue is the improper usage of the `pin!()` macro, for example: https://github.com/restatedev/restate/blob/bd7dc3ca75aae3852e9e77c2e07d1842bbbc50b9/crates/vqueues/src/scheduler/drr.rs#L793

Given a variable `x` with type `T`, the correct way of creating a value of type `Pin<&mut T>` is to write `pin!(x)` (and *not* `pin!(&mut x)`). If this value of type `Pin<&mut T>` needs to be used multiple times, it should be stored into a new local variable, and you should call the [`as_mut` method](https://doc.rust-lang.org/std/pin/struct.Pin.html#method.as_mut) each time you use the value.

Normally, the expression `pin!(&mut x)` would produce a value of type `Pin<&mut &mut T>`, which is probably not what you want. However, due to the bug in the Rust standard library, in some circumstances, the expression `pin!(&mut x)` may instead produce a value of type `Pin<&mut T>`. This breaks the invariant of `Pin`, and is therefore [unsound](https://doc.rust-lang.org/1.95.0/reference/behavior-considered-undefined.html#r-undefined.soundness) (may in some circumstances cause undefined behavior without any unsafe code). This bug will be fixed in a future version of Rust, and will unfortunately break your code.

For more details, see https://github.com/rust-lang/rust/issues/153438.

Sorry for the inconvenience.

### PR DESCRIPTION
**Pull request by AhmedSoliman**
I'm not loving this, but it works well enough to unblock.

Also fixes #4651

---
[//]: # (BEGIN SAPLING FOOTER)
Stack created with [Sapling](https://sapling-scm.com). Best reviewed with [ReviewStack](https://reviewstack.dev/restatedev/restate/pull/4642).
* #4635
* #4644
* #4649
* #4648
* #4647
* #4652
* #4646
* #4643
* __->__ #4642

### CODE REVIEW

File: crates/vqueues/src/scheduler/resource_manager/invoker_throttle.rs [hunk 1]
```diff
@@ -8,44 +8,405 @@
...
+        let Some(token_bucket) = self.token_bucket.as_ref() else {
+            return;
+        };
+
+        let waiters_in_window: Vec<_> = self
+            .waiters
+            .iter()
+            .copied()
+            .take(self.reservation_window)
+            .collect();
```
Comment (tillrohrmann) [inline crates/vqueues/src/scheduler/resource_manager/invoker_throttle.rs]: I know that this is a performance related comment so feel free to ignore: Would it make sense to have a reusable buffer for this instead of allocating the vec on every `replenish_reservations`?

Comment (tillrohrmann) [reply crates/vqueues/src/scheduler/resource_manager/invoker_throttle.rs]: If `self.reservations` is kept in sync with `self.waiters`, then we could theoretically skip the first `self.reservations.len()` and only take `self.reservation_window - self.reservations.len()`.

File: crates/vqueues/src/scheduler/resource_manager/invoker_throttle.rs [hunk 2]
```diff
@@ -8,44 +8,405 @@
...
+                    .timer_key
+                    .as_ref()
+                    .is_some_and(|queued_key| *queued_key == timer_key)
+            {
+                reservation.timer_key = None;
+            }
+        }
+    }
+
+    fn replenish_reservations(&mut self) {
```
Comment (tillrohrmann) [inline crates/vqueues/src/scheduler/resource_manager/invoker_throttle.rs:228]: I guess we are trying to replenish rather frequently for the sake of simplicity and avoiding corner cases, right?

File: crates/vqueues/src/scheduler/drr.rs [hunk 3]
```diff
@@ -836,6 +850,140 @@ mod tests {
...
+                db,
+                Concurrency::new_unlimited(),
+                Some(throttling_bucket),
+            )
+            .await,
+            db.clone(),
+            cache.view(),
+        );
+
+        let Poll::Ready(Ok(decision)) = poll_scheduler(pin!(&mut scheduler)) else {
```
Comment (tillrohrmann) [inline crates/vqueues/src/scheduler/drr.rs]: In preparation of Rust 1.97.0 should we rather use `Pin::new(&mut scheduler)` here?

File: crates/vqueues/src/scheduler/resource_manager.rs [hunk 4]
```diff
@@ -320,5 +324,12 @@ impl ResourceManager {
             );
             eligible.wake_up_queue(queue);
         }
+
+        while let Poll::Ready(Some(queue)) = self.invoker_throttling.poll_head(cx) {
+            tracing::trace!(
+                "waking up vqueue {queue:?} because invoker throttling token became available"
+            );
+            eligible.wake_up_queue(queue);
+        }
```
Comment (tillrohrmann) [inline crates/vqueues/src/scheduler/resource_manager.rs:333]: nit: Is there a point in looping `poll_head` of the invoker throttling? It looks as if there can only ever be a single ready head.

File: crates/vqueues/src/scheduler/resource_manager/invoker_throttle.rs [hunk 5]
```diff
@@ -8,44 +8,405 @@
...
+// to borrow from the token bucket, they would form a strand of lined-up waiters (in terms
+// of time delays, which is global to the node) and that would mean that we'd push vqueues
+// of other partitions on the same machine far in the future (after pre-booked strand of this
+// partition completes). This would mean that we'd cause unfairness between partitions and how
+// they share the global invoker throttling token bucket. To address this, we decide on a
+// somewhat arbitrary upper bound on the number of vqueues that can borrow from this partition
+// and leave the rest waiting without pre-booking. The downside is that for those vqueues outside
+// this bound, we'll not be able to determine their retry_at estimate (unless we add some kind
+// of heuristic, if needed).
+const MAX_RESERVED_WAITERS: usize = 17;
```
Comment (tillrohrmann) [inline crates/vqueues/src/scheduler/resource_manager/invoker_throttle.rs:36]: `17` is probably a magic number?

**Review by tillrohrmann — APPROVED** [review]
Thanks for creating this PR @AhmedSoliman. I think the `InvokerThrottlingLimiter` should do it's job. +1 for merging to unblock. If the limiter should become visible in profiles, we can invest in making it more efficient.

File: crates/vqueues/src/scheduler/resource_manager/invoker_throttle.rs (hunk 2 above)
Comment (AhmedSoliman) [reply crates/vqueues/src/scheduler/resource_manager/invoker_throttle.rs:228]: Yes, this is really a basic implementation to get us unblocked.

File: crates/vqueues/src/scheduler/resource_manager.rs (hunk 4 above)
Comment (AhmedSoliman) [reply crates/vqueues/src/scheduler/resource_manager.rs:333]: Not really, this is written with another implementation in mind. I'll leave it so that it accommodates the future implementation as well.

File: crates/vqueues/src/scheduler/resource_manager/invoker_throttle.rs (hunk 5 above)
Comment (AhmedSoliman) [reply crates/vqueues/src/scheduler/resource_manager/invoker_throttle.rs:36]: It absolutely is.

File: crates/vqueues/src/scheduler/resource_manager/invoker_throttle.rs (hunk 1 above)
Comment (AhmedSoliman) [reply crates/vqueues/src/scheduler/resource_manager/invoker_throttle.rs]: I wouldn't go far to try and optimize this, given that this is meant to be a stop-gap implementation.
